src/test/java/htsjdk/samtools/cram/build/SliceFactoryTest.java

```diff
@@ -0,0 +1,238 @@
+                // MULTI_REF-ness on the containing container, and they aren't efficient because they don't use
+                // refernce compression. But for non-cord sorted, we're likely to emit lots of MULTI_REF slices
+                // anyway (since we're less likely to accumulate a stream of reads mapped to the same contig).
+                //
+                // These tests are the cases that are handled differently for non-coord sorted inputs. Specifically,
+                // for non-coordinate sorted, when in multi-ref context, we continue to accumulate reads until we
+                // achieve a full slice of records, rather than emitting at MINIMUM_SINGLE_REFERENCE_SLICE_THRESHOLD.
+                { ReferenceContext.MULTIPLE_REFERENCE_ID, CRAMEncodingStrategy.DEFAULT_MINIMUM_SINGLE_REFERENCE_SLICE_THRESHOLD,
+                        MAPPED_REFERENCE_INDEX, ReferenceContext.MULTIPLE_REFERENCE_ID},
+                { ReferenceContext.MULTIPLE_REFERENCE_ID, CRAMEncodingStrategy.DEFAULT_MINIMUM_SINGLE_REFERENCE_SLICE_THRESHOLD,
+                        MAPPED_REFERENCE_INDEX + 1, ReferenceContext.MULTIPLE_REFERENCE_ID},
+                { ReferenceContext.MULTIPLE_REFERENCE_ID, CRAMEncodingStrategy.DEFAULT_MINIMUM_SINGLE_REFERENCE_SLICE_THRESHOLD,
+                        SAMRecord.NO_ALIGNMENT_REFERENCE_INDEX, ReferenceContext.MULTIPLE_REFERENCE_ID},
+                { ReferenceContext.MULTIPLE_REFERENCE_ID, CRAMEncodingStrategy.DEFAULT_READS_PER_SLICE - 1,
+                        MAPPED_REFERENCE_INDEX, ReferenceContext.MULTIPLE_REFERENCE_ID},
+                { ReferenceContext.MULTIPLE_REFERENCE_ID, CRAMEncodingStrategy.DEFAULT_READS_PER_SLICE - 1,
+                        MAPPED_REFERENCE_INDEX + 1, ReferenceContext.MULTIPLE_REFERENCE_ID},
+                { ReferenceContext.MULTIPLE_REFERENCE_ID, CRAMEncodingStrategy.DEFAULT_READS_PER_SLICE - 1,
+                        SAMRecord.NO_ALIGNMENT_REFERENCE_INDEX, ReferenceContext.MULTIPLE_REFERENCE_ID},
+        };
+    }
+
+    @Test(dataProvider = "emitSliceQuerynameSortedPositive")
+    private void testEmitSliceQuerynameSortedPositive(
+            final int currentReferenceContext,
+            final int nRecordsSeen,
+            final int nextReferenceContext,
+            final int expectedUpdatedReferenceContext) {
+        final SAMFileHeader querySortedSAMFileHeader = CRAMStructureTestHelper.SAM_FILE_HEADER.clone();
+        querySortedSAMFileHeader.setSortOrder(SAMFileHeader.SortOrder.queryname);
+        final SliceFactory sliceFactory = new SliceFactory(
+                new CRAMEncodingStrategy(),
+                CRAMStructureTestHelper.REFERENCE_SOURCE,
+                querySortedSAMFileHeader,
+                0L);
+        Assert.assertEquals(
+                sliceFactory.shouldEmitSlice(currentReferenceContext, nextReferenceContext, nRecordsSeen),
+                expectedUpdatedReferenceContext
+        );
+    }
+
```

Comment:
Are there any special negative test cases for the name-sorted case?

Reply:
I don't think there are.